moler/cmd/unix/top.py

```diff
@@ -0,0 +1,336 @@
+# -*- coding: utf-8 -*-
+"""
+Top command module.
+"""
+
+__author__ = 'Adrianna Pienkowska'
+__copyright__ = 'Copyright (C) 2018, Nokia'
+__email__ = 'adrianna.pienkowska@nokia.com'
+
+from moler.cmd.unix.genericunix import GenericUnixCommand
+from moler.exceptions import CommandFailure
+from moler.exceptions import ParsingDone
+import re
+
+
+class Top(GenericUnixCommand):
+    def __init__(self, connection, options=None, prompt=None, new_line_chars=None):
+        super(Top, self).__init__(connection=connection, prompt=prompt, new_line_chars=new_line_chars)
+        self.options = options
+        self._processes_list_headers = list()
+        self.current_ret = dict()
+
+    def build_command_string(self):
+        cmd = "top"
+        if self.options:
+            cmd = cmd + " " + self.options + " n 1"
+        else:
+            cmd = cmd + " " + "n 1"
+        return cmd
+
+    def on_new_line(self, line, is_full_line):
+        if is_full_line:
+            try:
+                self._command_failure(line)
+                self._parse_line(line)
+            except ParsingDone:
+                pass
+        return super(Top, self).on_new_line(line, is_full_line)
+
+    _re_error = re.compile(r'top:\s*(?P<ERROR_MSG>.*)')
+
+    def _command_failure(self, line):
+        if self._regex_helper.search_compiled(Top._re_error, line):
+            self.set_exception(CommandFailure(self, "ERROR: {}".format(self._regex_helper.group("ERROR_MSG"))))
+            raise ParsingDone
+
+    _re_top_row = re.compile(r'(?P<TOP_ROW>.*)\s-\s(?P<TIME>\d*:\d*:\d*)\sup\s(?P<UP_TIME>.*,.*),\s*(?P<USERS>.*)user'
+                             r'.*load average:(?P<LOAD_AVE> .*)')
+
+    _re_task_row = re.compile(r'(?P<TASK_ROW>Tasks):\s*(?P<TOTAL>\d*)\s*total,\s*(?P<RUN>\d*)\s*running,\s*'
+                              r'(?P<SLEEP>\d*)\s*sleeping,\s*(?P<STOP>\d*)\s*stopped, \s*(?P<ZOMBIE>\d*)\s*zombie')
+
+    _re_cpu_row = re.compile(r'.*(?P<CPU_ROW>Cpu).*:\s*(?P<US>\d*.\d*).*us,\s*(?P<SY>\d*.\d*).*sy,\s*(?P<NI>\d*.\d*)'
+                             r'.*ni,\s*(?P<ID>\d*.\d*).*id,\s*(?P<WA>\d*.\d*).*wa,\s*(?P<HI>\d*.\d*)'
+                             r'.*hi,\s*(?P<SI>\d*.\d*).*si,\s*(?P<ST>\d*.\d*).*st')
+
+    _re_memory_rows = re.compile(r'(?P<MEM>.*):\s*(?P<TOTAL_MEM>\d*)(?P<UNIT>.)\s*total,\s*(?P<FREE>\d*)\s*free,\s*'
+                                 r'(?P<USED>\d*)\s*used[,.]\s*(?P<OTHER>\d*)\s*')
+
+    _re_processes_header = re.compile(r'(?P<HEADER> .*PID.*)')
+
+    def _parse_line(self, line):
+        if self._regex_helper.search_compiled(Top._re_top_row, line):
```

Comment:
Please use common template for parsing in on_new_line.
Move parsing method into on_new_line.
Move checking regex into parsing method.